### DISCUSSION TITLE
Dev Conversation

### DISCUSSION THREAD

**Discussion by Daniele-rolli** [Daniele-rolli/Beaver-Notes#252] [General]
# 🛠️ Welcome to the Dev Conversation Thread!  

Hey everyone! This is the place to talk about all things **Beaver Notes** and **Beaver Pocket**,  from new features to improvements, or just cool ideas you’d love to see. Think of this as an open space for collaboration where devs, contributors, and the community can work together.

### 🚀 What This Thread Is For:  
- Talking about upcoming changes and improvements  
- Sharing ideas, feedback, and feature requests  
- Making decisions as a community  
- Connecting with developers, contributors, and other users  

### 💬 How to Join the Conversation:  
- **Got an idea?** Drop it here! The more context, examples, or mockups, the better.  
- **Have feedback?** We’re all ears!
- **Want to contribute code?** Check out [how to get involved](https://docs.beavernotes.com/beaver%20notes%20(dev)/2025/01/03/How-to-contribute.html).  
- **Be respectful and constructive.** Every idea matters, and we’re all working toward the same goal.  
- **This isn’t a replacement for GitHub issues** it’s more of a casual space to brainstorm, discuss, and stay in the loop.

**Comment by Daniele-rolli** [comment]
@bigshans I just discovered a new bug in your collapsible headings logic. When using footnotes, collapsing a heading also collapses the content under it, including the footnotes. However, when the heading is uncollapsed, its content is cleared, leaving only an empty list. If you do the same with a paragraph that contains references but no footnotes, the footnotes lose their content, but everything else still functions correctly.

edit:clarification 

https://github.com/user-attachments/assets/9cef4004-1983-4649-9150-e24ea2e06a56

**Reply by bigshans** [reply]
Not sure if I did something wrong, but I didn't manage to add footnotes. I rarely use footnotes, but I remember using the `[^1]` syntax to add footnotes automatically, but now it just adds references. I checked an older version and it does work this way, although there are some bugs.

**Reply by Daniele-rolli** [reply]
The issue you are facing might be partially caused by commit [75ec86aa486ab2f9206753dfe4c667bf09401257](https://github.com/Beaver-Notes/Beaver-Notes/commit/75ec86aa486ab2f9206753dfe4c667bf09401257), which introduced the following changes to fix issue #282:
```diff
- Document.extend({
-    content: 'block+ footnotes?',
- }),
+ Document,
```
It seems that the footnotes used in Beaver have been causing issues that went unnoticed for a while. It might be best to either replace the current footnotes extension or fix this before the next release.

**Reply by Daniele-rolli** [reply]
I reverted the issue caused by the removal of `content: 'block+ footnotes?'` with a change in commit b90401b. This fix seems to resolve both the gap cursor issues and the first issue shown in the video. Unfortunately, I haven't found a solution for the footnote content being erased when the heading is collapsed. Maybe you can look into it?

Here's the relevant change:
```diff
- Document,
+   Document.extend({
+     content: 'block+ (footnotes)?',
+     allowGapCursor: true,
+   }),
```

**Reply by bigshans** [reply]
I commented out the following code, and it seems to be back to normal. This code ignores the contents of the footnote when looking for the collapsed content, which seems to be why the footnote content disappears.

https://github.com/Beaver-Notes/Beaver-Notes/blob/b90401bf3220aab31faead8a1835a48cbc7569e3/packages/renderer/src/lib/tiptap/exts/collapse-heading/index.js#L57-L60

**Reply by Daniele-rolli** [reply]
I don't think that's the reason it disappears. That snippet actually prevents the footnotes list at the bottom from being collapsed along with the rest of the content. The actual reason might be that when a header is collapsed, the footnote reference is saved inside `collapsedContent`, but the actual footnote inside footnotes is not. So, when the heading is collapsed, the reference disappears, removing it from the footnotes list. Uncollapsing the heading adds the footnotes back because it restores the reference but doesn't add the content back.

**Reply by bigshans** [reply]
Do you want to collapse the footnote contents or just the contents other than the footnote?

**Reply by Daniele-rolli** [reply]
Honestly, I think we should collapse both since clicking on a footnote in a collapsed heading wouldn’t lead anywhere. That being said, my main concern right now is content retention rather than whether the footnote itself is collapsed. Thanks for the help, btw!

**Reply by bigshans** [reply]
I'm quite curious about the implementation of footnotes because I noticed that when I delete a reference, the footnote is also deleted. Our collapsing feature essentially works like a delete and insert operation. If the insertion after deletion doesn’t match the original, it naturally leads to significant issues. Moreover, I personally believe that references and footnotes should be decoupled, and footnotes and references shouldn’t be in a one-to-one relationship. Currently, if I insert a footnote 1 and then try to insert footnote 1 again, it turns into footnote 2, which is quite strange.

**Reply by Daniele-rolli** [reply]
I see where you're coming from, but having references and footnotes tied together is actually a good thing. It allows users to navigate seamlessly between a footnote and its reference. Plus, users can always type `[^1]` to add a footnote, and a number will be assigned automatically based on its position, this is much more convenient than having to remember the correct footnote number manually.  

I think decoupling them would ultimately hurt the UX. That being said, it would be ideal if we could find a way to handle this behavior in collapsible headings, for example, by ensuring that footnotes ignore the deletion of a reference in certain cases, or something along those lines.

**Reply by bigshans** [reply]
I am now storing the relevant footnotes under collapsible headings. I'm thinking about how to apply them when expanded and what to do when the collapse option is canceled. This will take some time. If you have a better solution, feel free to let me know. For now, I'll continue with my approach.

**Reply by bigshans** [reply]
Does this footnote have to be at the very bottom of the notes?

**Reply by Daniele-rolli** [reply]
No, I mean, it was designed that way. Why, what’s on your mind?

**Reply by bigshans** [reply]
I'm considering how the footnotes should be restored when the collapsible heading option is turned off. It would be much easier if the footnotes always remained at the bottom.

**Reply by Daniele-rolli** [reply]
Ok, so this would keep things as they are, with footnotes remaining at the bottom and references (aka. the numbers) staying within the paragraph.

**Comment by pranjal-jain-coder** [comment]
Hi,
I just wanted to suggest a small change. When a new note is created, it just gets added in the list of notes, but doesn't get opened in the editor which is kind of inconvenient. I think its probably an easy change.
Thanks

**Reply by Daniele-rolli** [reply]
How about we add an option in settings to disable this behavior instead of removing it completely?

**Comment by pranjal-jain-coder** [comment]
Hi,
I had a request for a feature to display word count.
Also, a detailed statistic like this would be nice
<img width="338" height="396" alt="Screenshot from 2025-09-11 16 27 43" src="https://github.com/user-attachments/assets/f60177ef-9fd0-4861-b3c0-f7aa4a4bf9d6" />

**Reply by Daniele-rolli** [reply]
Adding it to the roadmap
